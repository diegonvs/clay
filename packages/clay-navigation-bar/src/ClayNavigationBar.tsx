/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import classNames from 'classnames';
import React, {useEffect} from 'react';

import ClayButton from '@clayui/button';
import ClayIcon from '@clayui/icon';
import ClayLink from '@clayui/link';

interface Props extends React.HTMLAttributes<HTMLDivElement> {
	/**
	 * This property is used to pass a MouseEvent from a NavItem to a
	 * function when this element is clicked.
	 */
	onItemClicked?: (
		event: React.MouseEvent<HTMLLIElement, MouseEvent>
	) => void | undefined;

	/**
	 * The id that will be given to the Navigation<nav> element
	 */
	id?: string;

	/**
	 * An array of items that will be rendered on the dropdown.
	 */
	items: Array<{active?: boolean; href?: string; label: string}>;

	/**
	 * Determines the style of the Navigation Bar
	 */
	inverted?: boolean;

	/**
	 * Path to the location of the spritemap resource.
	 */
	spritemap: string;
}

const ClayNavigationBar: React.FunctionComponent<Props> = ({
	className,
	id,
	inverted = false,
	items,
	onItemClicked,
	spritemap,
	...otherProps
}) => {
	const [visible, setVisible] = React.useState(false);
	const [transitioning, setTransitioning] = React.useState(false);
	const contentRef = React.useRef<HTMLDivElement>(null);

	useEffect(() => {
		if (transitioning) {
			setCollapseHeight();
			if (visible) {
				removeCollapseHeight();
			}
		}
	}, [transitioning]);

	const handleTransitionEnd = (event: any) => {
		if (event.target === contentRef.current && transitioning && !visible) {
			setVisible(true);
			setTransitioning(false);
			removeCollapseHeight();
		} else if (event.target === contentRef.current) {
			setVisible(false);
			setTransitioning(false);
		}
	};

	const setCollapseHeight = () => {
		const elementCollapse = contentRef;
		if (elementCollapse && elementCollapse.current) {
			elementCollapse.current.setAttribute(
				'style',
				`height: ${elementCollapse.current.children[0].clientHeight}px`
			);
		}
	};

	const removeCollapseHeight = () => {
		if (contentRef && contentRef.current) {
			contentRef.current.style.removeProperty('height');
		}
	};

	const handleClickToggler = (event: any) => {
		event.preventDefault();
		if (visible && !transitioning) {
			setCollapseHeight();
		}

		if (!transitioning) {
			setTransitioning(true);
		}
	};

	return (
		<nav
			{...otherProps}
			className={classNames(
				className,
				'navbar',
				'navbar-collapse-absolute',
				'navbar-expand-md',
				'navbar-underline',
				'navigation-bar',
				{
					'navigation-bar-light': !inverted,
					'navigation-bar-secondary': inverted,
				}
			)}
			id={id}
		>
			<div className="container-fluid container-fluid-max-xl">
				<ClayLink
					className={classNames(
						'navbar-toggler',
						'navbar-toggler-link',
						{
							collapsed: !visible,
						}
					)}
					displayType="secondary"
					onClick={handleClickToggler}
				>
					{items.map(({active, label}) => {
						if (active) {
							return label;
						}
					})}
					<ClayIcon spritemap={spritemap} symbol="caret-bottom" />
				</ClayLink>

				<div
					className={classNames('navbar-collapse', {
						collapse: !transitioning,
						collapsing: transitioning,
						show: visible,
					})}
					onTransitionEnd={handleTransitionEnd}
					ref={contentRef}
				>
					<div className="container-fluid container-fluid-max-xl">
						<ul className="navbar-nav">
							{items.map(({active, href, label}, index) => (
								<li
									className={classNames('nav-item', {
										active,
									})}
									data-nav-item-index={index}
									key={index}
									onClick={onItemClicked}
								>
									{href ? (
										<ClayLink
											className={classNames('nav-link', {
												active,
											})}
											displayType="secondary"
											href={href}
										>
											<span className="navbar-text-truncate">
												{label}
											</span>
										</ClayLink>
									) : (
										<ClayButton
											block
											className={classNames('nav-link', {
												active,
											})}
											displayType="unstyled"
											small
										>
											<span className="navbar-text-truncate">
												{label}
											</span>
										</ClayButton>
									)}
								</li>
							))}
						</ul>
					</div>
				</div>
			</div>
		</nav>
	);
};

export default ClayNavigationBar;
