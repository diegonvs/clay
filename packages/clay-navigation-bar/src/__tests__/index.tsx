/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';

import ClayButton from '@clayui/button';
import ClayLink from '@clayui/link';
import ClayNavigationBar from '..';
import {
	act,
	fireEvent,
	getByTestId,
	getByText,
	render,
	waitForElement,
} from 'react-testing-library';

const spritemap = 'node_modules/clay-css/lib/images/icons/icons.svg';

describe('ClayNavigationBar', () => {
	it('renders', () => {
		const label = 'Item 1';

		const testRenderer = TestRenderer.create(
			<ClayNavigationBar
				inverted
				spritemap={spritemap}
				triggerLabel={label}
			>
				<ClayNavigationBar.Item active>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#1"
					>
						<span className="navbar-text-truncate">{label}</span>
					</ClayLink>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item>
					<ClayButton
						block
						className="nav-link"
						displayType="unstyled"
						small
					>
						<span className="navbar-text-truncate">{`Item 2`}</span>
					</ClayButton>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#3"
					>
						<span className="navbar-text-truncate">{`Item 3`}</span>
					</ClayLink>
				</ClayNavigationBar.Item>
			</ClayNavigationBar>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	// Skipped until React@16.9.0 will be released.
	// See https://github.com/liferay/clay/pull/1884#discussion_r282942724 for more informations
	xit('renders a dropdown when clicking the collapsed element from NavigationBar', async () => {
		const {container} = render(
			<ClayNavigationBar
				inverted
				spritemap={spritemap}
				triggerLabel={`Trigger Label`}
			>
				<ClayNavigationBar.Item active>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#1"
					>
						<span className="navbar-text-truncate">{`Item 1`}</span>
					</ClayLink>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item>
					<ClayButton
						block
						className="nav-link"
						displayType="unstyled"
						small
					>
						<span className="navbar-text-truncate">{`Item 2`}</span>
					</ClayButton>
				</ClayNavigationBar.Item>
			</ClayNavigationBar>
		);

		window.resizeTo(375, 812);

		fireEvent.click(getByText(container, 'Trigger Label'));
		fireEvent.transitionEnd(
			getByTestId(container, 'NavigationBarDropdown')
		);

		fireEvent.click(getByText(container, 'Trigger Label'));
		fireEvent.transitionEnd(
			getByTestId(container, 'NavigationBarDropdown')
		);

		let navigationBarDropdown;

		await act(async () => {
			navigationBarDropdown = await waitForElement(() =>
				getByTestId(container, 'NavigationBarDropdown')
			);
		});

		expect(navigationBarDropdown).toBeDefined();

		expect(navigationBarDropdown).toMatchSnapshot();
	});

	// Skipped until React@16.9.0 will be released.
	// See https://github.com/liferay/clay/pull/1884#discussion_r282942724 for more informations
	xit('collapses the dropdown expanded when trigger element is clicked', async () => {
		const {container} = render(
			<ClayNavigationBar
				inverted
				spritemap={spritemap}
				triggerLabel={`Trigger Label`}
			>
				<ClayNavigationBar.Item active>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#1"
					>
						<span className="navbar-text-truncate">{`Item 1`}</span>
					</ClayLink>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item>
					<ClayButton
						block
						className="nav-link"
						displayType="unstyled"
						small
					>
						<span className="navbar-text-truncate">{`Item 2`}</span>
					</ClayButton>
				</ClayNavigationBar.Item>
			</ClayNavigationBar>
		);

		window.resizeTo(375, 812);

		fireEvent.click(getByText(container, 'Trigger Label'));
		fireEvent.transitionEnd(
			getByTestId(container, 'NavigationBarDropdown')
		);

		fireEvent.click(getByText(container, 'Trigger Label'));
		fireEvent.transitionEnd(
			getByTestId(container, 'NavigationBarDropdown')
		);

		let navigationBarDropdown;

		await act(async () => {
			navigationBarDropdown = await waitForElement(() =>
				getByTestId(container, 'NavigationBarDropdown')
			);
		});

		expect(navigationBarDropdown).toBeDefined();

		expect(navigationBarDropdown).toMatchSnapshot();
	});

	it('renders when passing more than one active item', () => {
		const label = 'Item 1';

		const testRenderer = TestRenderer.create(
			<ClayNavigationBar
				inverted
				spritemap={spritemap}
				triggerLabel={label}
			>
				<ClayNavigationBar.Item active>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#1"
					>
						<span className="navbar-text-truncate">{label}</span>
					</ClayLink>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item active>
					<ClayButton
						block
						className="nav-link"
						displayType="unstyled"
						small
					>
						<span className="navbar-text-truncate">{`Item 2`}</span>
					</ClayButton>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#3"
					>
						<span className="navbar-text-truncate">{`Item 3`}</span>
					</ClayLink>
				</ClayNavigationBar.Item>
			</ClayNavigationBar>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should throw a warning when passing more than one active prop to child', () => {
		const mockWarnings = jest
			.spyOn(global.console, 'error')
			.mockImplementation(() => null);

		const label = 'Item 1';

		const testRenderer = TestRenderer.create(
			<ClayNavigationBar
				inverted
				spritemap={spritemap}
				triggerLabel={label}
			>
				<ClayNavigationBar.Item active>
					<ClayLink
						className="nav-link"
						displayType="secondary"
						href="#1"
					>
						<span className="navbar-text-truncate">{label}</span>
					</ClayLink>
				</ClayNavigationBar.Item>

				<ClayNavigationBar.Item active>
					<ClayButton
						block
						className="nav-link"
						displayType="unstyled"
						small
					>
						<span className="navbar-text-truncate">{`Item 2`}</span>
					</ClayButton>
				</ClayNavigationBar.Item>
			</ClayNavigationBar>
		);

		expect(mockWarnings).toBeCalled();
		expect(mockWarnings.mock.calls[0][0]).toBe(
			'Warning: ClayNavigationBar expects 0 or 1 active children, but received 2'
		);
		expect(testRenderer.toJSON).toMatchSnapshot();
		jest.resetAllMocks();
	});
});
