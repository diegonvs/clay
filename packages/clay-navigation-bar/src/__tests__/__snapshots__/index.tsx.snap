// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayNavigationBar collapses the dropdown expanded when trigger element is clicked 1`] = `
<div
  class="navbar-collapse collapse"
  data-testid="NavigationBarDropdown"
  style=""
>
  <div
    class="container-fluid container-fluid-max-xl"
  >
    <ul
      class="navbar-nav"
    >
      <li
        class="nav-item active"
      >
        <a
          class="nav-link nav-link active link-secondary"
          href="#1"
        >
          <span
            class="navbar-text-truncate"
          >
            Item 1
          </span>
        </a>
      </li>
      <li
        class="nav-item"
      >
        <button
          class="nav-link nav-link btn btn-unstyled btn-block btn-sm"
          type="button"
        >
          <span
            class="navbar-text-truncate"
          >
            Item 2
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`ClayNavigationBar renders 1`] = `
<nav
  className="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary"
>
  <div
    className="container-fluid container-fluid-max-xl"
  >
    <a
      aria-expanded={false}
      className="navbar-toggler navbar-toggler-link collapsed link-secondary"
      onClick={[Function]}
    >
      Item 1
      <svg
        className="lexicon-icon lexicon-icon-caret-bottom"
        role="presentation"
      >
        <use
          xlinkHref="node_modules/clay-css/lib/images/icons/icons.svg#caret-bottom"
        />
      </svg>
    </a>
    <div
      className="navbar-collapse collapse"
      data-testid="NavigationBarDropdown"
      onTransitionEnd={[Function]}
    >
      <div
        className="container-fluid container-fluid-max-xl"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link active link-secondary"
              href="#1"
            >
              <span
                className="navbar-text-truncate"
              >
                Item 1
              </span>
            </a>
          </li>
          <li
            className="nav-item"
          >
            <button
              className="nav-link btn btn-unstyled btn-block btn-sm"
              type="button"
            >
              <span
                className="navbar-text-truncate"
              >
                Item 2
              </span>
            </button>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link link-secondary"
              href="#3"
            >
              <span
                className="navbar-text-truncate"
              >
                Item 3
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar renders a dropdown when clicking the collapsed element from NavigationBar 1`] = `
<div
  class="navbar-collapse collapse"
  data-testid="NavigationBarDropdown"
  style=""
>
  <div
    class="container-fluid container-fluid-max-xl"
  >
    <ul
      class="navbar-nav"
    >
      <li
        class="nav-item active"
      >
        <a
          class="nav-link nav-link active link-secondary"
          href="#1"
        >
          <span
            class="navbar-text-truncate"
          >
            Item 1
          </span>
        </a>
      </li>
      <li
        class="nav-item"
      >
        <button
          class="nav-link nav-link btn btn-unstyled btn-block btn-sm"
          type="button"
        >
          <span
            class="navbar-text-truncate"
          >
            Item 2
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`ClayNavigationBar renders when passing more than one active item 1`] = `
<nav
  className="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary"
>
  <div
    className="container-fluid container-fluid-max-xl"
  >
    <a
      aria-expanded={false}
      className="navbar-toggler navbar-toggler-link collapsed link-secondary"
      onClick={[Function]}
    >
      Item 1
      <svg
        className="lexicon-icon lexicon-icon-caret-bottom"
        role="presentation"
      >
        <use
          xlinkHref="node_modules/clay-css/lib/images/icons/icons.svg#caret-bottom"
        />
      </svg>
    </a>
    <div
      className="navbar-collapse collapse"
      data-testid="NavigationBarDropdown"
      onTransitionEnd={[Function]}
    >
      <div
        className="container-fluid container-fluid-max-xl"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link active link-secondary"
              href="#1"
            >
              <span
                className="navbar-text-truncate"
              >
                Item 1
              </span>
            </a>
          </li>
          <li
            className="nav-item"
          >
            <button
              className="nav-link active btn btn-unstyled btn-block btn-sm"
              type="button"
            >
              <span
                className="navbar-text-truncate"
              >
                Item 2
              </span>
            </button>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link link-secondary"
              href="#3"
            >
              <span
                className="navbar-text-truncate"
              >
                Item 3
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;

exports[`ClayNavigationBar should throw a warning when passing more than one active prop to child 1`] = `[Function]`;
